Suggestion to upgrade to C# 9 is hiding true error

**Version Used**:
- _SDK:_ 6.0.100-preview.4.21255.9
- _PROJECT:_ NET 3.1

**Steps to Reproduce**:


```csharp
using System;

namespace ConsoleApp6
{
    class Program
    {
        static void Main(string[] args)
        {
            bool check1 = false;
            bool check2 = true;
            int value1 = 1;
            int value2 = 2;

            // ERROR HERE. These should be (int?)null instead of null
            TestMe(check1 ? null : value1, check2 ? null : value2);
        }

        static void TestMe(int? value1, int? value2)
        {

        }
    }
}

```

**Expected Behavior**:

The true error should be revealed:

> error CS0173: Type of conditional expression cannot be determined because there is no implicit conversion between '<null>' and 'int'

**Actual Behavior**:

Error is covered by a suggestion to upgrade to C# 9

> error CS8370: Feature 'target-typed conditional expression' is not available in C# 7.3. Please use language version 9.0 or greater.

--- a/src/Compilers/CSharp/Portable/Binder/Binder_Conversions.cs
+++ b/src/Compilers/CSharp/Portable/Binder/Binder_Conversions.cs
@@ -334,8 +334,16 @@ namespace Microsoft.CodeAnalysis.CSharp
                 : GenerateConversionForAssignment(destination, source.Alternative, diagnostics);
             var constantValue = FoldConditionalOperator(condition, trueExpr, falseExpr);
             hasErrors |= constantValue?.IsBad == true;
-            if (targetTyped && !destination.IsErrorType())
-                MessageID.IDS_FeatureTargetTypedConditional.CheckFeatureAvailability(diagnostics, source.Syntax);
+            if (targetTyped && !destination.IsErrorType() && !Compilation.IsFeatureEnabled(MessageID.IDS_FeatureTargetTypedConditional))
+            {
+                diagnostics.Add(
+                    ErrorCode.ERR_NoImplicitConvTargetTypedConditional,
+                    source.Syntax.Location,
+                    Compilation.LanguageVersion.ToDisplayString(),
+                    source.Consequence.Display,
+                    source.Alternative.Display,
+                    new CSharpRequiredLanguageVersion(MessageID.IDS_FeatureTargetTypedConditional.RequiredVersion()));
+            }
 
             return new BoundConditionalOperator(source.Syntax, isRef: false, condition, trueExpr, falseExpr, constantValue, source.Type, wasTargetTyped: targetTyped, destination, hasErrors)
                 .WithSuppression(source.IsSuppressed);

--- a/src/Compilers/CSharp/Portable/CSharpResources.resx
+++ b/src/Compilers/CSharp/Portable/CSharpResources.resx
@@ -6445,6 +6445,9 @@ To remove the warning, you can use /reference instead (set the Embed Interop Typ
   <data name="IDS_FeatureTargetTypedConditional" xml:space="preserve">
     <value>target-typed conditional expression</value>
   </data>
+  <data name="ERR_NoImplicitConvTargetTypedConditional" xml:space="preserve">
+    <value>Conditional expression is not valid in language version '{0}' because a common type was not found between '{1}' and '{2}'. To use a target-typed conversion, upgrade to language version '{3}' or greater.</value>
+  </data>
   <data name="ERR_DoesNotOverrideMethodFromObject" xml:space="preserve">
     <value>'{0}' does not override expected method from 'object'.</value>
   </data>

--- a/src/Compilers/CSharp/Portable/CSharpResources.resx
+++ b/src/Compilers/CSharp/Portable/CSharpResources.resx
@@ -6446,7 +6446,7 @@ To remove the warning, you can use /reference instead (set the Embed Interop Typ
     <value>target-typed conditional expression</value>
   </data>
   <data name="ERR_NoImplicitConvTargetTypedConditional" xml:space="preserve">
-    <value>Conditional expression is not valid in language version '{0}' because a common type was not found between '{1}' and '{2}'. To use a target-typed conversion, upgrade to language version '{3}' or greater.</value>
+    <value>Conditional expression is not valid in language version {0} because a common type was not found between '{1}' and '{2}'. To use a target-typed conversion, upgrade to language version {3} or greater.</value>
   </data>
   <data name="ERR_DoesNotOverrideMethodFromObject" xml:space="preserve">
     <value>'{0}' does not override expected method from 'object'.</value>

--- a/src/Compilers/CSharp/Portable/Errors/ErrorCode.cs
+++ b/src/Compilers/CSharp/Portable/Errors/ErrorCode.cs
@@ -1980,6 +1980,7 @@ namespace Microsoft.CodeAnalysis.CSharp
         ERR_MultipleFileScopedNamespace = 8954,
         ERR_FileScopedAndNormalNamespace = 8955,
         ERR_FileScopedNamespaceNotBeforeAllMembers = 8956,
+        ERR_NoImplicitConvTargetTypedConditional = 8957,
 
         #endregion
 

--- a/src/Compilers/CSharp/Portable/xlf/CSharpResources.cs.xlf
+++ b/src/Compilers/CSharp/Portable/xlf/CSharpResources.cs.xlf
@@ -823,8 +823,8 @@
         <note />
       </trans-unit>
       <trans-unit id="ERR_NoImplicitConvTargetTypedConditional">
-        <source>Conditional expression is not valid in language version '{0}' because a common type was not found between '{1}' and '{2}'. To use a target-typed conversion, upgrade to language version '{3}' or greater.</source>
-        <target state="new">Conditional expression is not valid in language version '{0}' because a common type was not found between '{1}' and '{2}'. To use a target-typed conversion, upgrade to language version '{3}' or greater.</target>
+        <source>Conditional expression is not valid in language version {0} because a common type was not found between '{1}' and '{2}'. To use a target-typed conversion, upgrade to language version {3} or greater.</source>
+        <target state="new">Conditional expression is not valid in language version {0} because a common type was not found between '{1}' and '{2}'. To use a target-typed conversion, upgrade to language version {3} or greater.</target>
         <note />
       </trans-unit>
       <trans-unit id="ERR_NoOutputDirectory">

--- a/src/Compilers/CSharp/Portable/xlf/CSharpResources.cs.xlf
+++ b/src/Compilers/CSharp/Portable/xlf/CSharpResources.cs.xlf
@@ -822,6 +822,11 @@
         <target state="translated">V základním typu {0} se nenašel žádný přístupný kopírovací konstruktor.</target>
         <note />
       </trans-unit>
+      <trans-unit id="ERR_NoImplicitConvTargetTypedConditional">
+        <source>Conditional expression is not valid in language version '{0}' because a common type was not found between '{1}' and '{2}'. To use a target-typed conversion, upgrade to language version '{3}' or greater.</source>
+        <target state="new">Conditional expression is not valid in language version '{0}' because a common type was not found between '{1}' and '{2}'. To use a target-typed conversion, upgrade to language version '{3}' or greater.</target>
+        <note />
+      </trans-unit>
       <trans-unit id="ERR_NoOutputDirectory">
         <source>Output directory could not be determined</source>
         <target state="translated">Nepovedlo se určit výstupní adresář.</target>

--- a/src/Compilers/CSharp/Portable/xlf/CSharpResources.de.xlf
+++ b/src/Compilers/CSharp/Portable/xlf/CSharpResources.de.xlf
@@ -822,6 +822,11 @@
         <target state="translated">Im Basistyp "{0}" wurde kein zugänglicher Kopierkonstruktor gefunden.</target>
         <note />
       </trans-unit>
+      <trans-unit id="ERR_NoImplicitConvTargetTypedConditional">
+        <source>Conditional expression is not valid in language version '{0}' because a common type was not found between '{1}' and '{2}'. To use a target-typed conversion, upgrade to language version '{3}' or greater.</source>
+        <target state="new">Conditional expression is not valid in language version '{0}' because a common type was not found between '{1}' and '{2}'. To use a target-typed conversion, upgrade to language version '{3}' or greater.</target>
+        <note />
+      </trans-unit>
       <trans-unit id="ERR_NoOutputDirectory">
         <source>Output directory could not be determined</source>
         <target state="translated">Das Ausgabeverzeichnis konnte nicht bestimmt werden.</target>

--- a/src/Compilers/CSharp/Portable/xlf/CSharpResources.de.xlf
+++ b/src/Compilers/CSharp/Portable/xlf/CSharpResources.de.xlf
@@ -823,8 +823,8 @@
         <note />
       </trans-unit>
       <trans-unit id="ERR_NoImplicitConvTargetTypedConditional">
-        <source>Conditional expression is not valid in language version '{0}' because a common type was not found between '{1}' and '{2}'. To use a target-typed conversion, upgrade to language version '{3}' or greater.</source>
-        <target state="new">Conditional expression is not valid in language version '{0}' because a common type was not found between '{1}' and '{2}'. To use a target-typed conversion, upgrade to language version '{3}' or greater.</target>
+        <source>Conditional expression is not valid in language version {0} because a common type was not found between '{1}' and '{2}'. To use a target-typed conversion, upgrade to language version {3} or greater.</source>
+        <target state="new">Conditional expression is not valid in language version {0} because a common type was not found between '{1}' and '{2}'. To use a target-typed conversion, upgrade to language version {3} or greater.</target>
         <note />
       </trans-unit>
       <trans-unit id="ERR_NoOutputDirectory">

--- a/src/Compilers/CSharp/Portable/xlf/CSharpResources.es.xlf
+++ b/src/Compilers/CSharp/Portable/xlf/CSharpResources.es.xlf
@@ -823,8 +823,8 @@
         <note />
       </trans-unit>
       <trans-unit id="ERR_NoImplicitConvTargetTypedConditional">
-        <source>Conditional expression is not valid in language version '{0}' because a common type was not found between '{1}' and '{2}'. To use a target-typed conversion, upgrade to language version '{3}' or greater.</source>
-        <target state="new">Conditional expression is not valid in language version '{0}' because a common type was not found between '{1}' and '{2}'. To use a target-typed conversion, upgrade to language version '{3}' or greater.</target>
+        <source>Conditional expression is not valid in language version {0} because a common type was not found between '{1}' and '{2}'. To use a target-typed conversion, upgrade to language version {3} or greater.</source>
+        <target state="new">Conditional expression is not valid in language version {0} because a common type was not found between '{1}' and '{2}'. To use a target-typed conversion, upgrade to language version {3} or greater.</target>
         <note />
       </trans-unit>
       <trans-unit id="ERR_NoOutputDirectory">

--- a/src/Compilers/CSharp/Portable/xlf/CSharpResources.es.xlf
+++ b/src/Compilers/CSharp/Portable/xlf/CSharpResources.es.xlf
@@ -822,6 +822,11 @@
         <target state="translated">No se encontró ningún constructor de copia accesible en el tipo de base "{0}".</target>
         <note />
       </trans-unit>
+      <trans-unit id="ERR_NoImplicitConvTargetTypedConditional">
+        <source>Conditional expression is not valid in language version '{0}' because a common type was not found between '{1}' and '{2}'. To use a target-typed conversion, upgrade to language version '{3}' or greater.</source>
+        <target state="new">Conditional expression is not valid in language version '{0}' because a common type was not found between '{1}' and '{2}'. To use a target-typed conversion, upgrade to language version '{3}' or greater.</target>
+        <note />
+      </trans-unit>
       <trans-unit id="ERR_NoOutputDirectory">
         <source>Output directory could not be determined</source>
         <target state="translated">No se pudo determinar el directorio de salida.</target>

--- a/src/Compilers/CSharp/Portable/xlf/CSharpResources.fr.xlf
+++ b/src/Compilers/CSharp/Portable/xlf/CSharpResources.fr.xlf
@@ -823,8 +823,8 @@
         <note />
       </trans-unit>
       <trans-unit id="ERR_NoImplicitConvTargetTypedConditional">
-        <source>Conditional expression is not valid in language version '{0}' because a common type was not found between '{1}' and '{2}'. To use a target-typed conversion, upgrade to language version '{3}' or greater.</source>
-        <target state="new">Conditional expression is not valid in language version '{0}' because a common type was not found between '{1}' and '{2}'. To use a target-typed conversion, upgrade to language version '{3}' or greater.</target>
+        <source>Conditional expression is not valid in language version {0} because a common type was not found between '{1}' and '{2}'. To use a target-typed conversion, upgrade to language version {3} or greater.</source>
+        <target state="new">Conditional expression is not valid in language version {0} because a common type was not found between '{1}' and '{2}'. To use a target-typed conversion, upgrade to language version {3} or greater.</target>
         <note />
       </trans-unit>
       <trans-unit id="ERR_NoOutputDirectory">

--- a/src/Compilers/CSharp/Portable/xlf/CSharpResources.fr.xlf
+++ b/src/Compilers/CSharp/Portable/xlf/CSharpResources.fr.xlf
@@ -822,6 +822,11 @@
         <target state="translated">Aucun constructeur de copie accessible n'a été trouvé dans le type de base '{0}'.</target>
         <note />
       </trans-unit>
+      <trans-unit id="ERR_NoImplicitConvTargetTypedConditional">
+        <source>Conditional expression is not valid in language version '{0}' because a common type was not found between '{1}' and '{2}'. To use a target-typed conversion, upgrade to language version '{3}' or greater.</source>
+        <target state="new">Conditional expression is not valid in language version '{0}' because a common type was not found between '{1}' and '{2}'. To use a target-typed conversion, upgrade to language version '{3}' or greater.</target>
+        <note />
+      </trans-unit>
       <trans-unit id="ERR_NoOutputDirectory">
         <source>Output directory could not be determined</source>
         <target state="translated">Impossible de déterminer le répertoire de sortie</target>

--- a/src/Compilers/CSharp/Portable/xlf/CSharpResources.it.xlf
+++ b/src/Compilers/CSharp/Portable/xlf/CSharpResources.it.xlf
@@ -822,6 +822,11 @@
         <target state="translated">Non è stato trovato alcun costruttore di copia accessibile nel tipo di base '{0}'.</target>
         <note />
       </trans-unit>
+      <trans-unit id="ERR_NoImplicitConvTargetTypedConditional">
+        <source>Conditional expression is not valid in language version '{0}' because a common type was not found between '{1}' and '{2}'. To use a target-typed conversion, upgrade to language version '{3}' or greater.</source>
+        <target state="new">Conditional expression is not valid in language version '{0}' because a common type was not found between '{1}' and '{2}'. To use a target-typed conversion, upgrade to language version '{3}' or greater.</target>
+        <note />
+      </trans-unit>
       <trans-unit id="ERR_NoOutputDirectory">
         <source>Output directory could not be determined</source>
         <target state="translated">Non è stato possibile individuare la directory di output</target>

--- a/src/Compilers/CSharp/Portable/xlf/CSharpResources.it.xlf
+++ b/src/Compilers/CSharp/Portable/xlf/CSharpResources.it.xlf
@@ -823,8 +823,8 @@
         <note />
       </trans-unit>
       <trans-unit id="ERR_NoImplicitConvTargetTypedConditional">
-        <source>Conditional expression is not valid in language version '{0}' because a common type was not found between '{1}' and '{2}'. To use a target-typed conversion, upgrade to language version '{3}' or greater.</source>
-        <target state="new">Conditional expression is not valid in language version '{0}' because a common type was not found between '{1}' and '{2}'. To use a target-typed conversion, upgrade to language version '{3}' or greater.</target>
+        <source>Conditional expression is not valid in language version {0} because a common type was not found between '{1}' and '{2}'. To use a target-typed conversion, upgrade to language version {3} or greater.</source>
+        <target state="new">Conditional expression is not valid in language version {0} because a common type was not found between '{1}' and '{2}'. To use a target-typed conversion, upgrade to language version {3} or greater.</target>
         <note />
       </trans-unit>
       <trans-unit id="ERR_NoOutputDirectory">

--- a/src/Compilers/CSharp/Portable/xlf/CSharpResources.ja.xlf
+++ b/src/Compilers/CSharp/Portable/xlf/CSharpResources.ja.xlf
@@ -822,6 +822,11 @@
         <target state="translated">基本型 '{0}' にアクセス可能なコピー コンストラクターが見つかりませんでした。</target>
         <note />
       </trans-unit>
+      <trans-unit id="ERR_NoImplicitConvTargetTypedConditional">
+        <source>Conditional expression is not valid in language version '{0}' because a common type was not found between '{1}' and '{2}'. To use a target-typed conversion, upgrade to language version '{3}' or greater.</source>
+        <target state="new">Conditional expression is not valid in language version '{0}' because a common type was not found between '{1}' and '{2}'. To use a target-typed conversion, upgrade to language version '{3}' or greater.</target>
+        <note />
+      </trans-unit>
       <trans-unit id="ERR_NoOutputDirectory">
         <source>Output directory could not be determined</source>
         <target state="translated">出力ディレクトリを特定できませんでした</target>

--- a/src/Compilers/CSharp/Portable/xlf/CSharpResources.ja.xlf
+++ b/src/Compilers/CSharp/Portable/xlf/CSharpResources.ja.xlf
@@ -823,8 +823,8 @@
         <note />
       </trans-unit>
       <trans-unit id="ERR_NoImplicitConvTargetTypedConditional">
-        <source>Conditional expression is not valid in language version '{0}' because a common type was not found between '{1}' and '{2}'. To use a target-typed conversion, upgrade to language version '{3}' or greater.</source>
-        <target state="new">Conditional expression is not valid in language version '{0}' because a common type was not found between '{1}' and '{2}'. To use a target-typed conversion, upgrade to language version '{3}' or greater.</target>
+        <source>Conditional expression is not valid in language version {0} because a common type was not found between '{1}' and '{2}'. To use a target-typed conversion, upgrade to language version {3} or greater.</source>
+        <target state="new">Conditional expression is not valid in language version {0} because a common type was not found between '{1}' and '{2}'. To use a target-typed conversion, upgrade to language version {3} or greater.</target>
         <note />
       </trans-unit>
       <trans-unit id="ERR_NoOutputDirectory">

--- a/src/Compilers/CSharp/Portable/xlf/CSharpResources.ko.xlf
+++ b/src/Compilers/CSharp/Portable/xlf/CSharpResources.ko.xlf
@@ -823,8 +823,8 @@
         <note />
       </trans-unit>
       <trans-unit id="ERR_NoImplicitConvTargetTypedConditional">
-        <source>Conditional expression is not valid in language version '{0}' because a common type was not found between '{1}' and '{2}'. To use a target-typed conversion, upgrade to language version '{3}' or greater.</source>
-        <target state="new">Conditional expression is not valid in language version '{0}' because a common type was not found between '{1}' and '{2}'. To use a target-typed conversion, upgrade to language version '{3}' or greater.</target>
+        <source>Conditional expression is not valid in language version {0} because a common type was not found between '{1}' and '{2}'. To use a target-typed conversion, upgrade to language version {3} or greater.</source>
+        <target state="new">Conditional expression is not valid in language version {0} because a common type was not found between '{1}' and '{2}'. To use a target-typed conversion, upgrade to language version {3} or greater.</target>
         <note />
       </trans-unit>
       <trans-unit id="ERR_NoOutputDirectory">

--- a/src/Compilers/CSharp/Portable/xlf/CSharpResources.ko.xlf
+++ b/src/Compilers/CSharp/Portable/xlf/CSharpResources.ko.xlf
@@ -822,6 +822,11 @@
         <target state="translated">기본 형식 '{0}'에 액세스 가능한 복사 생성자가 없습니다.</target>
         <note />
       </trans-unit>
+      <trans-unit id="ERR_NoImplicitConvTargetTypedConditional">
+        <source>Conditional expression is not valid in language version '{0}' because a common type was not found between '{1}' and '{2}'. To use a target-typed conversion, upgrade to language version '{3}' or greater.</source>
+        <target state="new">Conditional expression is not valid in language version '{0}' because a common type was not found between '{1}' and '{2}'. To use a target-typed conversion, upgrade to language version '{3}' or greater.</target>
+        <note />
+      </trans-unit>
       <trans-unit id="ERR_NoOutputDirectory">
         <source>Output directory could not be determined</source>
         <target state="translated">출력 디렉터리를 확인할 수 없습니다.</target>

--- a/src/Compilers/CSharp/Portable/xlf/CSharpResources.pl.xlf
+++ b/src/Compilers/CSharp/Portable/xlf/CSharpResources.pl.xlf
@@ -823,8 +823,8 @@
         <note />
       </trans-unit>
       <trans-unit id="ERR_NoImplicitConvTargetTypedConditional">
-        <source>Conditional expression is not valid in language version '{0}' because a common type was not found between '{1}' and '{2}'. To use a target-typed conversion, upgrade to language version '{3}' or greater.</source>
-        <target state="new">Conditional expression is not valid in language version '{0}' because a common type was not found between '{1}' and '{2}'. To use a target-typed conversion, upgrade to language version '{3}' or greater.</target>
+        <source>Conditional expression is not valid in language version {0} because a common type was not found between '{1}' and '{2}'. To use a target-typed conversion, upgrade to language version {3} or greater.</source>
+        <target state="new">Conditional expression is not valid in language version {0} because a common type was not found between '{1}' and '{2}'. To use a target-typed conversion, upgrade to language version {3} or greater.</target>
         <note />
       </trans-unit>
       <trans-unit id="ERR_NoOutputDirectory">

--- a/src/Compilers/CSharp/Portable/xlf/CSharpResources.pl.xlf
+++ b/src/Compilers/CSharp/Portable/xlf/CSharpResources.pl.xlf
@@ -822,6 +822,11 @@
         <target state="translated">W typie podstawowym „{0}” nie znaleziono dostępnego konstruktora kopiującego.</target>
         <note />
       </trans-unit>
+      <trans-unit id="ERR_NoImplicitConvTargetTypedConditional">
+        <source>Conditional expression is not valid in language version '{0}' because a common type was not found between '{1}' and '{2}'. To use a target-typed conversion, upgrade to language version '{3}' or greater.</source>
+        <target state="new">Conditional expression is not valid in language version '{0}' because a common type was not found between '{1}' and '{2}'. To use a target-typed conversion, upgrade to language version '{3}' or greater.</target>
+        <note />
+      </trans-unit>
       <trans-unit id="ERR_NoOutputDirectory">
         <source>Output directory could not be determined</source>
         <target state="translated">Nie można było określić katalogu wyjściowego</target>

--- a/src/Compilers/CSharp/Portable/xlf/CSharpResources.pt-BR.xlf
+++ b/src/Compilers/CSharp/Portable/xlf/CSharpResources.pt-BR.xlf
@@ -822,6 +822,11 @@
         <target state="translated">Não foi encontrado nenhum construtor de cópia acessível no tipo base '{0}'.</target>
         <note />
       </trans-unit>
+      <trans-unit id="ERR_NoImplicitConvTargetTypedConditional">
+        <source>Conditional expression is not valid in language version '{0}' because a common type was not found between '{1}' and '{2}'. To use a target-typed conversion, upgrade to language version '{3}' or greater.</source>
+        <target state="new">Conditional expression is not valid in language version '{0}' because a common type was not found between '{1}' and '{2}'. To use a target-typed conversion, upgrade to language version '{3}' or greater.</target>
+        <note />
+      </trans-unit>
       <trans-unit id="ERR_NoOutputDirectory">
         <source>Output directory could not be determined</source>
         <target state="translated">Não foi possível determinar o diretório de saída</target>

--- a/src/Compilers/CSharp/Portable/xlf/CSharpResources.pt-BR.xlf
+++ b/src/Compilers/CSharp/Portable/xlf/CSharpResources.pt-BR.xlf
@@ -823,8 +823,8 @@
         <note />
       </trans-unit>
       <trans-unit id="ERR_NoImplicitConvTargetTypedConditional">
-        <source>Conditional expression is not valid in language version '{0}' because a common type was not found between '{1}' and '{2}'. To use a target-typed conversion, upgrade to language version '{3}' or greater.</source>
-        <target state="new">Conditional expression is not valid in language version '{0}' because a common type was not found between '{1}' and '{2}'. To use a target-typed conversion, upgrade to language version '{3}' or greater.</target>
+        <source>Conditional expression is not valid in language version {0} because a common type was not found between '{1}' and '{2}'. To use a target-typed conversion, upgrade to language version {3} or greater.</source>
+        <target state="new">Conditional expression is not valid in language version {0} because a common type was not found between '{1}' and '{2}'. To use a target-typed conversion, upgrade to language version {3} or greater.</target>
         <note />
       </trans-unit>
       <trans-unit id="ERR_NoOutputDirectory">

--- a/src/Compilers/CSharp/Portable/xlf/CSharpResources.ru.xlf
+++ b/src/Compilers/CSharp/Portable/xlf/CSharpResources.ru.xlf
@@ -822,6 +822,11 @@
         <target state="translated">Доступный конструктор копий не найден в базовом типе "{0}".</target>
         <note />
       </trans-unit>
+      <trans-unit id="ERR_NoImplicitConvTargetTypedConditional">
+        <source>Conditional expression is not valid in language version '{0}' because a common type was not found between '{1}' and '{2}'. To use a target-typed conversion, upgrade to language version '{3}' or greater.</source>
+        <target state="new">Conditional expression is not valid in language version '{0}' because a common type was not found between '{1}' and '{2}'. To use a target-typed conversion, upgrade to language version '{3}' or greater.</target>
+        <note />
+      </trans-unit>
       <trans-unit id="ERR_NoOutputDirectory">
         <source>Output directory could not be determined</source>
         <target state="translated">Не удалось определить выходной каталог</target>

--- a/src/Compilers/CSharp/Portable/xlf/CSharpResources.ru.xlf
+++ b/src/Compilers/CSharp/Portable/xlf/CSharpResources.ru.xlf
@@ -823,8 +823,8 @@
         <note />
       </trans-unit>
       <trans-unit id="ERR_NoImplicitConvTargetTypedConditional">
-        <source>Conditional expression is not valid in language version '{0}' because a common type was not found between '{1}' and '{2}'. To use a target-typed conversion, upgrade to language version '{3}' or greater.</source>
-        <target state="new">Conditional expression is not valid in language version '{0}' because a common type was not found between '{1}' and '{2}'. To use a target-typed conversion, upgrade to language version '{3}' or greater.</target>
+        <source>Conditional expression is not valid in language version {0} because a common type was not found between '{1}' and '{2}'. To use a target-typed conversion, upgrade to language version {3} or greater.</source>
+        <target state="new">Conditional expression is not valid in language version {0} because a common type was not found between '{1}' and '{2}'. To use a target-typed conversion, upgrade to language version {3} or greater.</target>
         <note />
       </trans-unit>
       <trans-unit id="ERR_NoOutputDirectory">

--- a/src/Compilers/CSharp/Portable/xlf/CSharpResources.tr.xlf
+++ b/src/Compilers/CSharp/Portable/xlf/CSharpResources.tr.xlf
@@ -823,8 +823,8 @@
         <note />
       </trans-unit>
       <trans-unit id="ERR_NoImplicitConvTargetTypedConditional">
-        <source>Conditional expression is not valid in language version '{0}' because a common type was not found between '{1}' and '{2}'. To use a target-typed conversion, upgrade to language version '{3}' or greater.</source>
-        <target state="new">Conditional expression is not valid in language version '{0}' because a common type was not found between '{1}' and '{2}'. To use a target-typed conversion, upgrade to language version '{3}' or greater.</target>
+        <source>Conditional expression is not valid in language version {0} because a common type was not found between '{1}' and '{2}'. To use a target-typed conversion, upgrade to language version {3} or greater.</source>
+        <target state="new">Conditional expression is not valid in language version {0} because a common type was not found between '{1}' and '{2}'. To use a target-typed conversion, upgrade to language version {3} or greater.</target>
         <note />
       </trans-unit>
       <trans-unit id="ERR_NoOutputDirectory">

--- a/src/Compilers/CSharp/Portable/xlf/CSharpResources.tr.xlf
+++ b/src/Compilers/CSharp/Portable/xlf/CSharpResources.tr.xlf
@@ -822,6 +822,11 @@
         <target state="translated">'{0}' temel türünde erişilebilir kopya oluşturucu bulunamadı.</target>
         <note />
       </trans-unit>
+      <trans-unit id="ERR_NoImplicitConvTargetTypedConditional">
+        <source>Conditional expression is not valid in language version '{0}' because a common type was not found between '{1}' and '{2}'. To use a target-typed conversion, upgrade to language version '{3}' or greater.</source>
+        <target state="new">Conditional expression is not valid in language version '{0}' because a common type was not found between '{1}' and '{2}'. To use a target-typed conversion, upgrade to language version '{3}' or greater.</target>
+        <note />
+      </trans-unit>
       <trans-unit id="ERR_NoOutputDirectory">
         <source>Output directory could not be determined</source>
         <target state="translated">Çıkış dizini belirlenemedi</target>

--- a/src/Compilers/CSharp/Portable/xlf/CSharpResources.zh-Hans.xlf
+++ b/src/Compilers/CSharp/Portable/xlf/CSharpResources.zh-Hans.xlf
@@ -822,6 +822,11 @@
         <target state="translated">在基类型“{0}”中找不到可访问的复制构造函数。</target>
         <note />
       </trans-unit>
+      <trans-unit id="ERR_NoImplicitConvTargetTypedConditional">
+        <source>Conditional expression is not valid in language version '{0}' because a common type was not found between '{1}' and '{2}'. To use a target-typed conversion, upgrade to language version '{3}' or greater.</source>
+        <target state="new">Conditional expression is not valid in language version '{0}' because a common type was not found between '{1}' and '{2}'. To use a target-typed conversion, upgrade to language version '{3}' or greater.</target>
+        <note />
+      </trans-unit>
       <trans-unit id="ERR_NoOutputDirectory">
         <source>Output directory could not be determined</source>
         <target state="translated">无法确定输出目录</target>

--- a/src/Compilers/CSharp/Portable/xlf/CSharpResources.zh-Hans.xlf
+++ b/src/Compilers/CSharp/Portable/xlf/CSharpResources.zh-Hans.xlf
@@ -823,8 +823,8 @@
         <note />
       </trans-unit>
       <trans-unit id="ERR_NoImplicitConvTargetTypedConditional">
-        <source>Conditional expression is not valid in language version '{0}' because a common type was not found between '{1}' and '{2}'. To use a target-typed conversion, upgrade to language version '{3}' or greater.</source>
-        <target state="new">Conditional expression is not valid in language version '{0}' because a common type was not found between '{1}' and '{2}'. To use a target-typed conversion, upgrade to language version '{3}' or greater.</target>
+        <source>Conditional expression is not valid in language version {0} because a common type was not found between '{1}' and '{2}'. To use a target-typed conversion, upgrade to language version {3} or greater.</source>
+        <target state="new">Conditional expression is not valid in language version {0} because a common type was not found between '{1}' and '{2}'. To use a target-typed conversion, upgrade to language version {3} or greater.</target>
         <note />
       </trans-unit>
       <trans-unit id="ERR_NoOutputDirectory">

--- a/src/Compilers/CSharp/Portable/xlf/CSharpResources.zh-Hant.xlf
+++ b/src/Compilers/CSharp/Portable/xlf/CSharpResources.zh-Hant.xlf
@@ -822,6 +822,11 @@
         <target state="translated">在基底類型 '{0}' 中找不到可存取的複製建構函式。</target>
         <note />
       </trans-unit>
+      <trans-unit id="ERR_NoImplicitConvTargetTypedConditional">
+        <source>Conditional expression is not valid in language version '{0}' because a common type was not found between '{1}' and '{2}'. To use a target-typed conversion, upgrade to language version '{3}' or greater.</source>
+        <target state="new">Conditional expression is not valid in language version '{0}' because a common type was not found between '{1}' and '{2}'. To use a target-typed conversion, upgrade to language version '{3}' or greater.</target>
+        <note />
+      </trans-unit>
       <trans-unit id="ERR_NoOutputDirectory">
         <source>Output directory could not be determined</source>
         <target state="translated">無法判斷輸出目錄</target>

--- a/src/Compilers/CSharp/Portable/xlf/CSharpResources.zh-Hant.xlf
+++ b/src/Compilers/CSharp/Portable/xlf/CSharpResources.zh-Hant.xlf
@@ -823,8 +823,8 @@
         <note />
       </trans-unit>
       <trans-unit id="ERR_NoImplicitConvTargetTypedConditional">
-        <source>Conditional expression is not valid in language version '{0}' because a common type was not found between '{1}' and '{2}'. To use a target-typed conversion, upgrade to language version '{3}' or greater.</source>
-        <target state="new">Conditional expression is not valid in language version '{0}' because a common type was not found between '{1}' and '{2}'. To use a target-typed conversion, upgrade to language version '{3}' or greater.</target>
+        <source>Conditional expression is not valid in language version {0} because a common type was not found between '{1}' and '{2}'. To use a target-typed conversion, upgrade to language version {3} or greater.</source>
+        <target state="new">Conditional expression is not valid in language version {0} because a common type was not found between '{1}' and '{2}'. To use a target-typed conversion, upgrade to language version {3} or greater.</target>
         <note />
       </trans-unit>
       <trans-unit id="ERR_NoOutputDirectory">

--- a/src/Compilers/CSharp/Test/IOperation/IOperation/IOperationTests_IThrowOperation.cs
+++ b/src/Compilers/CSharp/Test/IOperation/IOperation/IOperationTests_IThrowOperation.cs
@@ -2171,7 +2171,7 @@ class C
             var compilation = CreateCompilation(source, parseOptions: TestOptions.Regular8);
 
             compilation.VerifyDiagnostics(
-                // (6,13): error CS8957: Conditional expression is not valid in language version '8.0' because a common type was not found between '<throw expression>' and '<throw expression>'. To use a target-typed conversion, upgrade to language version '9.0' or greater.
+                // (6,13): error CS8957: Conditional expression is not valid in language version 8.0 because a common type was not found between '<throw expression>' and '<throw expression>'. To use a target-typed conversion, upgrade to language version 9.0 or greater.
                 //         x = y ? throw ex1 : throw ex2;
                 Diagnostic(ErrorCode.ERR_NoImplicitConvTargetTypedConditional, "y ? throw ex1 : throw ex2").WithArguments("8.0", "<throw expression>", "<throw expression>", "9.0").WithLocation(6, 13)
                 );

--- a/src/Compilers/CSharp/Test/IOperation/IOperation/IOperationTests_IThrowOperation.cs
+++ b/src/Compilers/CSharp/Test/IOperation/IOperation/IOperationTests_IThrowOperation.cs
@@ -2171,9 +2171,9 @@ class C
             var compilation = CreateCompilation(source, parseOptions: TestOptions.Regular8);
 
             compilation.VerifyDiagnostics(
-                // (6,13): error CS8400: Feature 'target-typed conditional expression' is not available in C# 8.0. Please use language version 9.0 or greater.
+                // (6,13): error CS8957: Conditional expression is not valid in language version '8.0' because a common type was not found between '<throw expression>' and '<throw expression>'. To use a target-typed conversion, upgrade to language version '9.0' or greater.
                 //         x = y ? throw ex1 : throw ex2;
-                Diagnostic(ErrorCode.ERR_FeatureNotAvailableInVersion8, "y ? throw ex1 : throw ex2").WithArguments("target-typed conditional expression", "9.0").WithLocation(6, 13)
+                Diagnostic(ErrorCode.ERR_NoImplicitConvTargetTypedConditional, "y ? throw ex1 : throw ex2").WithArguments("8.0", "<throw expression>", "<throw expression>", "9.0").WithLocation(6, 13)
                 );
 
             string expectedOperationTree = @"

--- a/src/Compilers/CSharp/Test/Semantic/Semantics/SemanticErrorTests.cs
+++ b/src/Compilers/CSharp/Test/Semantic/Semantics/SemanticErrorTests.cs
@@ -6531,7 +6531,7 @@ public class Square
                 // (21,16): error CS0172: Type of conditional expression cannot be determined because 'Square.Circle' and 'Square' implicitly convert to one another
                 //       var o1 = (1 == 1) ? aa : ii;   // CS0172
                 Diagnostic(ErrorCode.ERR_AmbigQM, "(1 == 1) ? aa : ii").WithArguments("Square.Circle", "Square").WithLocation(21, 16),
-                // (22,19): error CS8957: Conditional expression is not valid in language version '8.0' because a common type was not found between 'Square.Circle' and 'Square'. To use a target-typed conversion, upgrade to language version '9.0' or greater.
+                // (22,19): error CS8957: Conditional expression is not valid in language version 8.0 because a common type was not found between 'Square.Circle' and 'Square'. To use a target-typed conversion, upgrade to language version 9.0 or greater.
                 //       object o2 = (1 == 1) ? aa : ii;   // CS8652
                 Diagnostic(ErrorCode.ERR_NoImplicitConvTargetTypedConditional, "(1 == 1) ? aa : ii").WithArguments("8.0", "Square.Circle", "Square", "9.0").WithLocation(22, 19)
                 );

--- a/src/Compilers/CSharp/Test/Semantic/Semantics/SemanticErrorTests.cs
+++ b/src/Compilers/CSharp/Test/Semantic/Semantics/SemanticErrorTests.cs
@@ -6531,9 +6531,9 @@ public class Square
                 // (21,16): error CS0172: Type of conditional expression cannot be determined because 'Square.Circle' and 'Square' implicitly convert to one another
                 //       var o1 = (1 == 1) ? aa : ii;   // CS0172
                 Diagnostic(ErrorCode.ERR_AmbigQM, "(1 == 1) ? aa : ii").WithArguments("Square.Circle", "Square").WithLocation(21, 16),
-                // (22,19): error CS8400: Feature 'target-typed conditional expression' is not available in C# 8.0. Please use language version 9.0 or greater.
+                // (22,19): error CS8957: Conditional expression is not valid in language version '8.0' because a common type was not found between 'Square.Circle' and 'Square'. To use a target-typed conversion, upgrade to language version '9.0' or greater.
                 //       object o2 = (1 == 1) ? aa : ii;   // CS8652
-                Diagnostic(ErrorCode.ERR_FeatureNotAvailableInVersion8, "(1 == 1) ? aa : ii").WithArguments("target-typed conditional expression", "9.0").WithLocation(22, 19)
+                Diagnostic(ErrorCode.ERR_NoImplicitConvTargetTypedConditional, "(1 == 1) ? aa : ii").WithArguments("8.0", "Square.Circle", "Square", "9.0").WithLocation(22, 19)
                 );
         }
 

--- a/src/Compilers/CSharp/Test/Semantic/Semantics/TargetTypedConditionalOperatorTests.cs
+++ b/src/Compilers/CSharp/Test/Semantic/Semantics/TargetTypedConditionalOperatorTests.cs
@@ -640,9 +640,10 @@ class Program
                 // (15,16): error CS0029: Cannot implicitly convert type 'A' to 'B'
                 //         return (b ? a : 0) + a;
                 Diagnostic(ErrorCode.ERR_NoImplicitConv, "(b ? a : 0) + a").WithArguments("A", "B").WithLocation(15, 16),
-                // (15,17): error CS8370: Feature 'target-typed conditional expression' is not available in C# 7.3. Please use language version 9.0 or greater.
+                // (15,17): error CS8960: Type of conditional expression cannot be determined because a common type was not found between 'A' and 'int'. To use a target-typed conversion, upgrade to language version '9.0' or greater.
                 //         return (b ? a : 0) + a;
-                Diagnostic(ErrorCode.ERR_FeatureNotAvailableInVersion7_3, "b ? a : 0").WithArguments("target-typed conditional expression", "9.0").WithLocation(15, 17));
+                Diagnostic(ErrorCode.ERR_NoImplicitConvTargetTypedConditional, "b ? a : 0").WithArguments("A", "int", "9.0").WithLocation(15, 17));
+
             CreateCompilation(source, parseOptions: TestOptions.Regular.WithLanguageVersion(MessageID.IDS_FeatureTargetTypedConditional.RequiredVersion())).VerifyDiagnostics(
                 // (15,16): error CS0029: Cannot implicitly convert type 'A' to 'B'
                 //         return (b ? a : 0) + a;
@@ -689,6 +690,28 @@ class Program
             var typeInfo = model.GetTypeInfo(expr);
             Assert.Null(typeInfo.Type);
             Assert.Equal("System.Int32?", typeInfo.ConvertedType.ToTestDisplayString());
+        }
+
+        [Fact]
+        public void DiagnosticClarity_LangVersion8()
+        {
+            var source = @"
+class C
+{
+    void M(bool b)
+    {
+        int? i = b ? 1 : null;
+    }
+}
+";
+            var comp = CreateCompilation(source, parseOptions: TestOptions.Regular8);
+            comp.VerifyDiagnostics(
+                // (6,18): error CS8957: Conditional expression is not valid in language version '8.0' because a common type was not found between 'int' and '<null>'. To use a target-typed conversion, upgrade to language version '9.0' or greater.
+                //         int? i = b ? 1 : null;
+                Diagnostic(ErrorCode.ERR_NoImplicitConvTargetTypedConditional, "b ? 1 : null").WithArguments("8.0", "int", "<null>", "9.0").WithLocation(6, 18));
+
+            comp = CreateCompilation(source);
+            comp.VerifyDiagnostics();
         }
     }
 }

--- a/src/Compilers/CSharp/Test/Semantic/Semantics/TargetTypedConditionalOperatorTests.cs
+++ b/src/Compilers/CSharp/Test/Semantic/Semantics/TargetTypedConditionalOperatorTests.cs
@@ -640,9 +640,9 @@ class Program
                 // (15,16): error CS0029: Cannot implicitly convert type 'A' to 'B'
                 //         return (b ? a : 0) + a;
                 Diagnostic(ErrorCode.ERR_NoImplicitConv, "(b ? a : 0) + a").WithArguments("A", "B").WithLocation(15, 16),
-                // (15,17): error CS8960: Type of conditional expression cannot be determined because a common type was not found between 'A' and 'int'. To use a target-typed conversion, upgrade to language version '9.0' or greater.
+                // (15,17): error CS8957: Conditional expression is not valid in language version 7.3 because a common type was not found between 'A' and 'int'. To use a target-typed conversion, upgrade to language version 9.0 or greater.
                 //         return (b ? a : 0) + a;
-                Diagnostic(ErrorCode.ERR_NoImplicitConvTargetTypedConditional, "b ? a : 0").WithArguments("A", "int", "9.0").WithLocation(15, 17));
+                Diagnostic(ErrorCode.ERR_NoImplicitConvTargetTypedConditional, "b ? a : 0").WithArguments("7.3", "A", "int", "9.0").WithLocation(15, 17));
 
             CreateCompilation(source, parseOptions: TestOptions.Regular.WithLanguageVersion(MessageID.IDS_FeatureTargetTypedConditional.RequiredVersion())).VerifyDiagnostics(
                 // (15,16): error CS0029: Cannot implicitly convert type 'A' to 'B'
@@ -706,7 +706,7 @@ class C
 ";
             var comp = CreateCompilation(source, parseOptions: TestOptions.Regular8);
             comp.VerifyDiagnostics(
-                // (6,18): error CS8957: Conditional expression is not valid in language version '8.0' because a common type was not found between 'int' and '<null>'. To use a target-typed conversion, upgrade to language version '9.0' or greater.
+                // (6,18): error CS8957: Conditional expression is not valid in language version 8.0 because a common type was not found between 'int' and '<null>'. To use a target-typed conversion, upgrade to language version 9.0 or greater.
                 //         int? i = b ? 1 : null;
                 Diagnostic(ErrorCode.ERR_NoImplicitConvTargetTypedConditional, "b ? 1 : null").WithArguments("8.0", "int", "<null>", "9.0").WithLocation(6, 18));
 

--- a/src/EditorFeatures/CSharpTest/Diagnostics/UpgradeProject/UpgradeProjectTests.cs
+++ b/src/EditorFeatures/CSharpTest/Diagnostics/UpgradeProject/UpgradeProjectTests.cs
@@ -1071,5 +1071,20 @@ class C1 : [|I1|]
                 expected: LanguageVersion.CSharp10,
                 new CSharpParseOptions(LanguageVersion.CSharp9));
         }
+
+        [Fact]
+        public async Task UpgradeProjectForTargetTypedConditional()
+        {
+            await TestLanguageVersionUpgradedAsync(@"
+class C
+{
+    void M(bool b)
+    {
+        int? i = [|b ? 1 : null|];
+    }
+}",
+                expected: LanguageVersion.CSharp9,
+                new CSharpParseOptions(LanguageVersion.CSharp8));
+        }
     }
 }

--- a/src/Features/CSharp/Portable/UpgradeProject/CSharpUpgradeProjectCodeFixProvider.cs
+++ b/src/Features/CSharp/Portable/UpgradeProject/CSharpUpgradeProjectCodeFixProvider.cs
@@ -49,6 +49,7 @@ namespace Microsoft.CodeAnalysis.CSharp.UpgradeProject
                 "CS8904", // error CS8904: Invalid variance: The type parameter 'T1' must be contravariantly valid on 'I2<T1, T2>.M1(T1)' unless language version 'preview' or greater is used. 'T1' is covariant.
                 "CS8912", // error CS8912: Inheriting from a record with a sealed 'Object.ToString' is not supported in C# {0}. Please use language version '{1}' or greater.
                 "CS8704", // error CS8704: 'Test1' does not implement interface member 'I1.M1()'. 'Test1.M1()' cannot implicitly implement a non-public member in C# 9.0. Please use language version 'preview' or greater.
+                "CS8957", // error CS8957: Conditional expression is not valid in language version '8.0' because a common type was not found between 'int' and '<null>'. To use a target-typed conversion, upgrade to language version '9.0' or greater.
             });
 
         public override string UpgradeThisProjectResource => CSharpFeaturesResources.Upgrade_this_project_to_csharp_language_version_0;

